### ISSUE TITLE
Where to place loading of CUDAService?

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#28575]
PR #28537 introduces `CUDAService` that needs to be loaded in any configuration that may include CUDA EDModules. The service works also on machines without GPUs, and is used to communicate that fact to elsewhere in CMSSW.

The question is then where exactly the `process.load("HeterogeneousCore.CUDAServices.CUDAService_cfi")` should be placed? In the patatrack fork we modified `ConfigBuilder` to always include it in the `cmsDriver.py`-generated configuration files.

An alternative could be to import `CUDAService` into all cfi/cff files that define CUDA EDModules. On the other hand, we don't tend to do that with other services (like `DQMStore` or `RandomNumberGeneratorService`).

Requiring the framework to always load `CUDAService` internally does not sound a good solution either (at this stage at least).

Are there any better ideas?

**Comment by fabiocos** [comment]
assign operations,core

**Comment by fabiocos** [comment]
@makortel so practically what you did was to update https://github.com/cms-sw/cmssw/blob/master/Configuration/StandardSequences/python/Services_cff.py ? This sounds as a natural entry point, unless you want to have it conditional to some specific flag

**Comment by makortel** [comment]
@fabiocos I put a `process.load()` in `ConfigBuilder`
https://github.com/cms-sw/cmssw/blob/692e725b916209cb65ca5a3135ecfe9a6719253f/Configuration/Applications/python/ConfigBuilder.py#L912-L925

Thanks for pointing out the `Configuration.StandardSequences.Services_cff`, it indeed looks a better placement. The intention indeed is that `CUDAService` could be loaded by any job, and the service internally handles the case of no GPUs.

**Comment by makortel** [comment]
Based on the discussion yesterday in the core meeting (given #28576), probably it would make sense (as an interim solution) to load `CUDAService_cfi` only if the `gpu` modifier (#28522) is given to the `cmsDriver.py`. Therefore the `loadAndRemember()` call would stay in the `ConfigBuilder`, protected with an "if gpu". @fabiocos, @fwyzard, what do you think?

**Comment by fwyzard** [comment]
I understood that `if gpu` is deprecated... apart from that, I'm fine with this approach :-)

**Comment by makortel** [comment]
The actual line would not be `if gpu.isChosen()`, but along "if `gpu` in `--processModifiers`".

**Comment by makortel** [comment]
I agree it is not pretty, but would limit the set of broken workflows in `ppc64le` only to the ones really using CUDA.

**Comment by makortel** [comment]
Thinking further
```python
def loadCUDAService(process):
    process.load("HeterogeneousCore.CUDAServices.CUDAService_cff")
modifyConfigurationStandardSequencesServicesCUDAService = gpu.makeProcessModifier(loadCUDAService)
```
in `Configuration/StandardSequences/Services_cff` should work as well and wouldn't require touching `ConfigBuilder`.

**Comment by fwyzard** [comment]
Look clean enough to me ...

**Comment by fabiocos** [comment]
@makortel from a quick test I did

```
from Configuration.ProcessModifiers.gpu_cff import *
def loadCUDAService(process):
    process.load("HeterogeneousCore.CUDAServices.CUDAService_cfi")
modifyConfigurationStandardSequencesServicesCUDAService_ = gpu.makeProcessModifier(loadCUDAService)
```

is correctly uploaded in case the process modifier ```gpu``` is part of the Era
